### DISCUSSION TITLE
Additional Responses in OpenAPI and Mypy

### DISCUSSION THREAD

**Discussion by angely-dev** [fastapi/fastapi#12056] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class Message(BaseModel):
    message: str


responses = {
    403: {"model": Message, "description": "Not enough privileges"},
    404: {"model": Message, "description": "Item not found"},
}


@app.get("/items/{item_id}", responses={**responses})
def read_item(item_id: int):
    pass
```

### Description

**I don't think FastAPI is the root cause of the problem. I think it has to do with how Mypy handles dictionary unpacking. But I'd be glad the view of those familiar with Mypy.**

As per [Additional Responses in OpenAPI](https://fastapi.tiangolo.com/advanced/additional-responses/#combine-predefined-responses-and-custom-ones), I often use this pattern:

```py
responses = {
    403: {"model": Message, "description": "Not enough privileges"},
    404: {"model": Message, "description": "Item not found"},
}

@app.get("/endpoint1/{item_id}", responses={**responses})
@app.get("/endpoint2/{item_id}", responses={**responses})
@app.get("/endpoint3/{item_id}", responses={**responses})
```

Which is very convenient, as it allows to reuse same error responses across multiple endpoints.

**Problem is: it doesn't pass Mypy checking.**

```sh
$ mypy api.py
api.py:17: error: Unpacked dict entry 0 has incompatible type "dict[int, dict[str, object]]"; expected "SupportsKeysAndGetItem[int | str, dict[str, Any]]"  [dict-item]
Found 1 error in 1 file (checked 1 source file)
```

As a workaround, I am using:

```py
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class Message(BaseModel):
    message: str


e403 = {"model": Message, "description": "Not enough privileges"}
e404 = {"model": Message, "description": "Item not found"}


@app.get("/items/{item_id}", responses={403: e403, 404: e404})
def read_item(item_id: int):
    pass
```

Which indeed passes Mypy checking:

```sh
$ mypy api.py 
Success: no issues found in 1 source file
```

But it is less convenient to use.

Are you aware of a better solution? Without ignoring faulty lines.

Thanks!

### Operating System

Linux

### FastAPI Version

0.112.1

### Pydantic Version

2.8.2

### Python Version

3.10.12

**Comment by YuriiMotov** [comment, marked as answer]
You can solve this by adding the explicit type annotation to `responses`:

```python
from typing import Any
from pydantic import BaseModel

from fastapi import FastAPI


app = FastAPI()


class Message(BaseModel):
    message: str


responses: dict[int | str, dict[str, Any]] = {
    403: {"model": Message, "description": "Not enough privileges"},
    404: {"model": Message, "description": "Item not found"},
}


@app.get("/endpoint1/{item_id}", responses={**responses})
async def endpoint1(item_id: int):
    return {"message": str(item_id)}

```

**Reply by angely-dev** [reply]
Thanks for pointing out the solution :)
